kafka-connect-s3/src/test/java/io/confluent/connect/s3/DataWriterParquetTest.java

```diff
@@ -506,6 +532,205 @@ public void testProjectNoVersion() throws Exception {
+      for (int optionalItems = 0; optionalItems < 2; ++optionalItems) {
+        for (int mapRegular = 0; mapRegular < 2; ++mapRegular) {
+          for (int mapOptional = 0; mapOptional < 2; ++mapOptional) {
+            for (int hasNested = 0; hasNested < 2; ++hasNested) {
+              for (int hasNestedArray = 0; hasNestedArray < 2; ++hasNestedArray) {
+                SchemaConfig conf = new SchemaConfig();
+                SchemaConfig nested = null, nestedArray = null;
+                if (hasNested != 0) {
+                    // Invert tests so as to hit in isolation
+                    nested = new SchemaConfig(
+                        "nested_schema",
+                        regularItems == 0,
+                        optionalItems == 0,
+                        mapRegular == 0,
+                        mapOptional == 0,
+                        /*nested=*/ null,
+                        /*nestedArray=*/null
+                    );
+                }
+                if (hasNestedArray != 0) {
+                  // Invert tests so as to hit in isolation
+                  nestedArray = new SchemaConfig(
+                      "array_of_schema",
+                      regularItems == 0,
+                      optionalItems == 0,
+                      mapRegular == 0,
+                      mapOptional == 0,
+                      /*nested=*/ null,
+                      /*nestedArray=*/null
+                  );
+                }
+                conf.nested = new SchemaConfig(
+                    "nested_schema",
+                    regularItems != 0,
+                    optionalItems != 0,
+                    mapRegular != 0,
+                    mapOptional != 0,
+                    nested,
+                    nestedArray
+                );
+                Schema schema = conf.create();
+                if (schema.fields().size() == 0) {
+                  // Base case of everything is false
+                  continue;
+                }
+                final boolean hasArrayOptional =
+                    ParquetRecordWriterProvider.schemaHasArrayOfOptionalItems(
+                      schema,
+                      /*seenSchemas=*/null
+                  );
+                final boolean shouldHaveArrayOptional = conf.hasOptionalItems();
+                assertEquals(hasArrayOptional, shouldHaveArrayOptional);
+              }
+            }
+          }
+        }
+      }
+    }
+  }
+
+  private List<SinkRecord> createRecordsWithArrayOfOptionalString(
+      int size,
+      Set<TopicPartition> partitions
+  ) {
+    SchemaConfig conf = new SchemaConfig();
+    conf.regularItems = true;
+    conf.optionalItems = true;
+    conf.nested = new SchemaConfig(
+        "nested_schema",
+        /*regularItems=*/true,
+        /*optionalItems=*/true,
+        /*mapRegular=*/false,
+        /*mapOptional=*/false,
+        /*nested=*/null,
+        /*nestedArray=*/null
```

Comment:
IMO I don't think we should commit to such a coding convention. This is not currently part of the style of the connector and modern IDEs offer enough visibility to which argument a value points to while reading the code.

Reply:
ok, but I think it's hard to review on github, stuff like 
```java
  callFunction(true, false, true, true, true, null, false, null);
```